File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        getRightVisible(root, result, 0);
        return result;
    }

    private void getRightVisible(TreeNode node, List<Integer> result, int level) {
        if(node == null) return;
        if(result.size() == level) {
            result.add(node.val);
        }
        if(node.right != null){
            getRightVisible(node.right, result, level + 1);
        }
        if (node.left != null) {
            getRightVisible(node.left, result, level + 1);
        }
    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun buildTree(preorder: IntArray, inorder: IntArray): TreeNode? {
        if(preorder.isEmpty()) return null
        if(preorder.size == 1) return TreeNode(preorder[0])

        val node = TreeNode(preorder[0])

        val inLeft : IntArray = inorder.slice(0..<inorder.indexOf(preorder[0])).toIntArray()
        val inRight : IntArray = inorder.slice(inorder.indexOf(preorder[0])+1..inorder.size-1).toIntArray()

        var index = 1
        while (index < preorder.size) {
            if(!inLeft.contains(preorder[index])){
                break
            }
            index++
        }

        val preLeft = preorder.slice(1..index-1).toIntArray()
        val preRight = preorder.slice(index..<preorder.size).toIntArray()

        node.left = buildTree(preLeft, inLeft)
        node.right = buildTree(preRight, inRight)
        return node
    }
}
